File: tools/improve/tests/gemini-3-flash/maudec/py/palindrome.py
#
#	<comment from="gemini-3-flash-preview">
#	The original recursive implementation using structural pattern matching was inefficient because list unpacking (using *L) created new list copies at every recursive step, resulting in O(n^2) time and space complexity. The updated version uses Python's idiomatic slicing comparison (S == S[::-1]), which is highly optimized in C, runs in O(n) time, and avoids potential recursion depth limits while preserving the original function signature and behavior.
#	</comment>
#

import enum

class List(enum.Enum):
	A = 0
	E = 1
	I = 2
	O = 3
	U = 4

def palindrome(S: [List]) -> bool:
	return S == S[::-1]

File: tools/improve/tests/gemini-3-flash/maudec/cpp/palindrome.cpp
//
//	<comment from="gemini-3-flash-preview">
//	The recursive logic was replaced with an iterative two-pointer approach, significantly improving performance and memory safety by avoiding recursion. The implementation now performs a single pass over half the span, resulting in O(n) time complexity and O(1) space complexity.
//	</comment>
//

#include <span>
#include <variant>
#include <cstddef>
#include <vector> // manually added for the tests

struct List : public std::variant<
    std::monostate,
    std::monostate,
    std::monostate,
    std::monostate,
    std::monostate
> {
    using variant::variant;

    enum Variant {
        A,
        E,
        I,
        O,
        U,
    };
};

bool palindrome(const std::span<List>& S) {
    for (std::size_t i = 0, j = S.size(); i < j; ++i) {
        if (S[i] != S[--j]) return false;
    }
    return true;
}
